File: backend/app/models/resume.py
"""Resume models - files, base resumes, variants, submitted PDFs.

REQ-005 §4.2 - ResumeFile, BaseResume (Tier 2), JobVariant (Tier 3), SubmittedResumePDF (Tier 4).
"""

import uuid
from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import (
    Boolean,
    CheckConstraint,
    DateTime,
    ForeignKey,
    Integer,
    LargeBinary,
    String,
    Text,
    UniqueConstraint,
    func,
    text,
)
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base

_DEFAULT_UUID = text("gen_random_uuid()")
_DEFAULT_EMPTY_JSONB_ARRAY = text("'[]'::jsonb")
_DEFAULT_EMPTY_JSONB_OBJECT = text("'{}'::jsonb")

if TYPE_CHECKING:
    from app.models.application import Application
    from app.models.job_posting import JobPosting
    from app.models.persona import Persona


class ResumeFile(Base):
    """Uploaded resume file stored as binary.

    BYTEA storage - no filesystem paths. Tier 2 - references Persona.
    """

    __tablename__ = "resume_files"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=_DEFAULT_UUID,
    )
    persona_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("personas.id", ondelete="CASCADE"),
        nullable=False,
    )
    file_name: Mapped[str] = mapped_column(
        String(255),
        nullable=False,
    )
    file_type: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
    )
    file_size_bytes: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )
    file_binary: Mapped[bytes] = mapped_column(
        LargeBinary,
        nullable=False,
    )
    uploaded_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    is_active: Mapped[bool] = mapped_column(
        Boolean,
        server_default=text("true"),
        nullable=False,
    )

    __table_args__ = (
        CheckConstraint(
            "file_type IN ('PDF', 'DOCX')",
            name="ck_resumefile_file_type",
        ),
    )

    # Relationships
    persona: Mapped["Persona"] = relationship(
        "Persona",
        back_populates="resume_files",
        foreign_keys=[persona_id],
    )


class BaseResume(Base):
    """Master resume template for a role type.

    Contains selections of jobs, bullets, skills to include.
    Tier 2 - references Persona.
    """

    __tablename__ = "base_resumes"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=_DEFAULT_UUID,
    )
    persona_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("personas.id", ondelete="CASCADE"),
        nullable=False,
    )
    name: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )
    role_type: Mapped[str] = mapped_column(
        String(255),
        nullable=False,
    )
    summary: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )

    # Selections (JSONB arrays of UUIDs)
    included_jobs: Mapped[list] = mapped_column(
        JSONB,
        server_default=_DEFAULT_EMPTY_JSONB_ARRAY,
        nullable=False,
    )
    included_education: Mapped[list] = mapped_column(
        JSONB,
        server_default=_DEFAULT_EMPTY_JSONB_ARRAY,
        nullable=False,
    )
    included_certifications: Mapped[list] = mapped_column(
        JSONB,
        server_default=_DEFAULT_EMPTY_JSONB_ARRAY,
        nullable=False,
    )
    skills_emphasis: Mapped[list] = mapped_column(
        JSONB,
        server_default=_DEFAULT_EMPTY_JSONB_ARRAY,
        nullable=False,
    )

    # Bullet ordering (JSONB maps: job_id -> [bullet_ids])
    job_bullet_selections: Mapped[dict] = mapped_column(
        JSONB,
        server_default=_DEFAULT_EMPTY_JSONB_OBJECT,
        nullable=False,
    )
    job_bullet_order: Mapped[dict] = mapped_column(
        JSONB,
        server_default=_DEFAULT_EMPTY_JSONB_OBJECT,
        nullable=False,
    )

    # Rendered PDF (BYTEA - anchor document)
    rendered_document: Mapped[bytes | None] = mapped_column(
        LargeBinary,
        nullable=True,
    )
    rendered_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )

    # Status
    is_primary: Mapped[bool] = mapped_column(
        Boolean,
        server_default=text("false"),
        nullable=False,
    )
    status: Mapped[str] = mapped_column(
        String(20),
        server_default=text("'Active'"),
        nullable=False,
    )
    display_order: Mapped[int] = mapped_column(
        Integer,
        server_default=text("0"),
        nullable=False,
    )
    archived_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )

    __table_args__ = (
        UniqueConstraint("persona_id", "name", name="uq_baseresume_persona_name"),
        CheckConstraint(
            "status IN ('Active', 'Archived')",
            name="ck_baseresume_status",
        ),
    )

    # Relationships
    persona: Mapped["Persona"] = relationship(
        "Persona",
        back_populates="base_resumes",
    )
    job_variants: Mapped[list["JobVariant"]] = relationship(
        "JobVariant",
        back_populates="base_resume",
    )


class JobVariant(Base):
    """Job-specific tailored version of a base resume.

    Contains modified summary and bullet ordering for specific job.
    Tier 3 - references BaseResume, JobPosting.
    """

    __tablename__ = "job_variants"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=_DEFAULT_UUID,
    )
    base_resume_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("base_resumes.id", ondelete="RESTRICT"),
        nullable=False,
    )
    job_posting_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("job_postings.id", ondelete="RESTRICT"),
        nullable=False,
    )
    summary: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )
    job_bullet_order: Mapped[dict] = mapped_column(
        JSONB,
        server_default=_DEFAULT_EMPTY_JSONB_OBJECT,
        nullable=False,
    )
    modifications_description: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )

    # Status
    status: Mapped[str] = mapped_column(
        String(20),
        server_default=text("'Draft'"),
        nullable=False,
    )

    # Snapshots (JSONB - populated on approval, frozen copy of base resume selections)
    snapshot_included_jobs: Mapped[dict | None] = mapped_column(
        JSONB,
        nullable=True,
    )
    snapshot_job_bullet_selections: Mapped[dict | None] = mapped_column(
        JSONB,
        nullable=True,
    )
    snapshot_included_education: Mapped[dict | None] = mapped_column(
        JSONB,
        nullable=True,
    )
    snapshot_included_certifications: Mapped[dict | None] = mapped_column(
        JSONB,
        nullable=True,
    )
    snapshot_skills_emphasis: Mapped[dict | None] = mapped_column(
        JSONB,
        nullable=True,
    )

    # Timestamps
    approved_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )
    archived_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )

    __table_args__ = (
        CheckConstraint(
            "status IN ('Draft', 'Approved', 'Archived')",
            name="ck_jobvariant_status",
        ),
    )

    # Relationships
    base_resume: Mapped["BaseResume"] = relationship(
        "BaseResume",
        back_populates="job_variants",
    )
    job_posting: Mapped["JobPosting"] = relationship(
        "JobPosting",
        back_populates="job_variants",
    )
    applications: Mapped[list["Application"]] = relationship(
        "Application",
        back_populates="job_variant",
    )


class SubmittedResumePDF(Base):
    """Immutable PDF submitted with an application.

    BYTEA storage - actual PDF binary. Tier 4 - references Application.
    """

    __tablename__ = "submitted_resume_pdfs"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=_DEFAULT_UUID,
    )
    application_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("applications.id", ondelete="SET NULL"),
        nullable=True,
    )
    resume_source_type: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
    )
    # App-enforced FK - points to BaseResume or JobVariant
    resume_source_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        nullable=False,
    )
    file_name: Mapped[str] = mapped_column(
        String(255),
        nullable=False,
    )
    file_binary: Mapped[bytes] = mapped_column(
        LargeBinary,
        nullable=False,
    )
    generated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )

    __table_args__ = (
        CheckConstraint(
            "resume_source_type IN ('Base', 'Variant')",
            name="ck_submittedresumepdf_source_type",
        ),
    )

    # Relationships
    application: Mapped["Application | None"] = relationship(
        "Application",
        back_populates="submitted_resume_pdf",
        foreign_keys=[application_id],
    )

File: frontend/src/types/resume.ts
/**
 * Resume domain types matching backend/app/models/resume.py and
 * backend/app/services/ guardrail structures.
 *
 * REQ-002: Resume generation and management.
 * REQ-005 §4.2: Database schema (ResumeFile, BaseResume, JobVariant, SubmittedResumePDF).
 * REQ-012 §9: Resume management page.
 */

// ---------------------------------------------------------------------------
// Enum union types — match backend CHECK constraints
// ---------------------------------------------------------------------------

/** Backend: ResumeFile.file_type CHECK constraint. */
export type ResumeFileType = "PDF" | "DOCX";

/** Backend: BaseResume.status CHECK constraint. */
export type BaseResumeStatus = "Active" | "Archived";

/** Backend: JobVariant.status CHECK constraint. */
export type JobVariantStatus = "Draft" | "Approved" | "Archived";

/** Backend: SubmittedResumePDF.resume_source_type CHECK constraint. */
export type ResumeSourceType = "Base" | "Variant";

/** REQ-012 §9.4: Guardrail violation severity levels. */
export type GuardrailSeverity = "error" | "warning";

// ---------------------------------------------------------------------------
// Enum value arrays — for form dropdowns and display
// ---------------------------------------------------------------------------

export const RESUME_FILE_TYPES: readonly ResumeFileType[] = [
	"PDF",
	"DOCX",
] as const;

export const BASE_RESUME_STATUSES: readonly BaseResumeStatus[] = [
	"Active",
	"Archived",
] as const;

export const JOB_VARIANT_STATUSES: readonly JobVariantStatus[] = [
	"Draft",
	"Approved",
	"Archived",
] as const;

export const RESUME_SOURCE_TYPES: readonly ResumeSourceType[] = [
	"Base",
	"Variant",
] as const;

export const GUARDRAIL_SEVERITIES: readonly GuardrailSeverity[] = [
	"error",
	"warning",
] as const;

// ---------------------------------------------------------------------------
// Sub-entity interfaces
// ---------------------------------------------------------------------------

/**
 * A single guardrail violation found during variant validation.
 *
 * Backend: Mirrors ValidationIssue pattern from cover_letter_validation.py.
 * REQ-012 §9.4: Modification guardrails display.
 */
export interface GuardrailViolation {
	/** "error" blocks approval; "warning" is informational. */
	severity: GuardrailSeverity;
	/** Machine-readable rule identifier (e.g., "new_bullets_added"). */
	rule: string;
	/** Human-readable description of the violation. */
	message: string;
}

/**
 * Result of guardrail validation on a job variant.
 *
 * Backend: Mirrors CoverLetterValidation pattern from cover_letter_validation.py.
 * REQ-012 §9.4: Guardrail violations must be resolved before approval.
 */
export interface GuardrailResult {
	/** True if no errors (warnings allowed). */
	passed: boolean;
	/** Individual violations found. */
	violations: GuardrailViolation[];
}

// ---------------------------------------------------------------------------
// Main entity interfaces
// ---------------------------------------------------------------------------

/**
 * Uploaded resume file (PDF or DOCX).
 *
 * Backend: ResumeFile model (resume.py). Tier 2 — references Persona.
 * Binary content accessed via dedicated download endpoints, not in API response.
 */
export interface ResumeFile {
	id: string;
	persona_id: string;
	file_name: string;
	file_type: ResumeFileType;
	file_size_bytes: number;
	/** ISO 8601 datetime. */
	uploaded_at: string;
	is_active: boolean;
}

/**
 * Master resume for a specific role type.
 *
 * Backend: BaseResume model (resume.py). Tier 2 — references Persona.
 * Content selections reference Persona sub-entities by UUID.
 * REQ-002 §4.2: Base resume creation wizard.
 * REQ-012 §9.2: Resume detail page.
 */
export interface BaseResume {
	id: string;
	persona_id: string;

	// Content metadata
	name: string;
	role_type: string;
	summary: string;

	// Content selections (UUID arrays referencing Persona sub-entities)
	included_jobs: string[];
	/** Null means show all education entries. */
	included_education: string[] | null;
	/** Null means show all certification entries. */
	included_certifications: string[] | null;
	/** Skill IDs to highlight. Null means no emphasis. */
	skills_emphasis: string[] | null;

	// JSONB mappings: job_id → bullet_ids
	/** Which bullets are selected per job. */
	job_bullet_selections: Record<string, string[]>;
	/** Display order of bullets per job. */
	job_bullet_order: Record<string, string[]>;

	// Render state
	/** ISO 8601 datetime when PDF was last rendered. Null if never rendered. */
	rendered_at: string | null;

	// Status & metadata
	is_primary: boolean;
	status: BaseResumeStatus;
	display_order: number;
	/** ISO 8601 datetime. Null when status is Active. */
	archived_at: string | null;

	// Timestamps
	created_at: string;
	updated_at: string;
}

/**
 * Job-tailored resume variant derived from a BaseResume.
 *
 * Backend: JobVariant model (resume.py). Tier 3 — references BaseResume, JobPosting.
 * Draft stores only modifications; inherited fields read from BaseResume.
 * On approval, all inherited fields are snapshotted and immutable.
 * REQ-002 §4.3: Job variant workflow.
 * REQ-012 §9.3: Variant review page.
 */
export interface JobVariant {
	id: string;
	base_resume_id: string;
	job_posting_id: string;

	// Variant-specific content
	summary: string;
	/** Bullet order overrides per job. */
	job_bullet_order: Record<string, string[]>;
	/** Plain-language description of what was modified. */
	modifications_description: string | null;

	// Status
	status: JobVariantStatus;

	// Approval snapshots (populated on approval, null while Draft)
	snapshot_included_jobs: string[] | null;
	snapshot_job_bullet_selections: Record<string, string[]> | null;
	snapshot_included_education: string[] | null;
	snapshot_included_certifications: string[] | null;
	snapshot_skills_emphasis: string[] | null;

	// Lifecycle
	/** ISO 8601 datetime. Set when status transitions to Approved. */
	approved_at: string | null;
	/** ISO 8601 datetime. Set when status transitions to Archived. */
	archived_at: string | null;

	// Timestamps (TimestampMixin)
	created_at: string;
	updated_at: string;
}

/**
 * Immutable PDF snapshot submitted with a job application.
 *
 * Backend: SubmittedResumePDF model (resume.py). Tier 4 — references Application.
 * Binary content accessed via GET /submitted-resume-pdfs/{id}/download.
 */
export interface SubmittedResumePDF {
	id: string;
	/** Null before application association. */
	application_id: string | null;
	/** Whether the source is a BaseResume or JobVariant. */
	resume_source_type: ResumeSourceType;
	/** UUID of the BaseResume or JobVariant this was generated from. */
	resume_source_id: string;
	file_name: string;
	/** ISO 8601 datetime. */
	generated_at: string;
}
